packages/features/calendar-sync/calendarSubscription.service.ts

```diff
@@ -0,0 +1,261 @@
+import logger from "@calcom/lib/logger";
+import { safeStringify } from "@calcom/lib/safeStringify";
+import prisma from "@calcom/prisma";
+import { CalendarSubscriptionStatus } from "@calcom/prisma/enums";
+
+import { CalendarSubscriptionRepository } from "./calendarSubscription.repository";
+
+const log = logger.getSubLogger({ prefix: ["[CalendarSubscriptionService]"] });
+type Timestamp = number;
+type ProviderSubscriptionDetails = {
+  id: string | null;
+  kind: string | null;
+  resourceId: string | null;
+  resourceUri: string | null;
+  expiration: Timestamp | null;
+};
+/**
+ * CalendarSubscriptionService centralizes subscription management logic across calendar providers
+ * It handles finding, reusing, and creating calendar subscriptions
+ */
+export class CalendarSubscriptionService {
+  /**
+   * Find an existing subscription from either CalendarSubscription or SelectedCalendar tables
+   * Returns normalized subscription data regardless of source
+   */
+  static async findActiveProviderSubscriptionInSelectedCalendarToo({
+    externalCalendarId,
+    integration,
+  }: {
+    externalCalendarId: string;
+    integration: string;
+  }): Promise<
+    | ((
+        | {
+            source: "CalendarSubscription";
+            subscriptionId: string;
+            fromSelectedCalendar: false;
+          }
+        | {
+            source: "SelectedCalendar";
+            selectedCalendarId: string;
+            fromSelectedCalendar: true;
+          }
+      ) & {
+        providerDetails: ProviderSubscriptionDetails;
+      })
+    | null
+  > {
+    log.debug("Finding existing subscription", safeStringify({ externalCalendarId, integration }));
+
+    // First check for active subscription in CalendarSubscription
+    const existingSubscription = await CalendarSubscriptionRepository.findFirst({
```

Comment:
Is there any way this could be a "findUnique"? As @hbjORbj discovered this can have up to 90% performance improvement when we have multiple records.

Reply:
@hariombalhara

Reply:
@zomars I don't see an easy way to make it _findUnique_. The problem is that CalendarSubscription also has _credentialId_ which allows different credentials to be there that are connected to same externalCalendarId. We don't have an easy way to identify which credentialId to be used among all such credentialIds. 


Also, because the query is called via cron, maybe performance improvement won't matter much here? Open to ideas.